### DISCUSSION TITLE
How to update members of parameter hierarchy?

### DISCUSSION THREAD

**Discussion by EricDallAgnol** [atoti/atoti#581] [Q&A] [answered]
Hi there,

I'm looking for advice on implementing an Analysis Hierarchy within Atoti.

I want to create my members from values read from a csv, for that I know `create_parameter_hierarchy_from_members` will do the job.
But my concern is about keeping my members up-to-date. Indeed, each day a new input file for my analysis hierarchy is loaded, and I might need to remove/add some members.

Currently, with Java AP, it was handled by a custom class implementing `ISelectionListener` attached to the isolated store upon which the  analysis hierarchy is created. As a result, as soon as a new transaction on this store is detected (ie a new input file loaded), we are triggering the `buildDiscriminatorPaths()` if an update is needed.
Is there something similar possible with Atoti, or another way to implement this logic ?

Many thanks,

**Comment by bloussou** [comment, marked as answer]
Hello,
`create_parameter_hierarchy_from_members()` is creating a table behind the scene and you can access it. I will show it using the [`create_parameter_hierarchy_from_members()` documentation](https://docs.atoti.io/latest/lib/atoti/atoti.cube.html#atoti.Cube.create_parameter_hierarchy_from_members):
```python
df = pd.DataFrame(
    {
        "Seller": ["Seller_1", "Seller_2", "Seller_3"],
        "Prices": [
            [2.5, 49.99, 3.0, 54.99],
            [2.6, 50.99, 2.8, 57.99],
            [2.99, 44.99, 3.6, 59.99],
        ],
    }
)

table = session.read_pandas(df, table_name="Seller prices")
cube = session.create_cube(table)
l, m = cube.levels, cube.measures

cube.create_parameter_hierarchy_from_members(
    "ProductID",
    ["aBk3", "ceJ4", "aBk5", "ceJ9"],
    index_measure_name="Product index",
)

m["Prices"] = tt.value(table["Prices"])
m["Product price"] = m["Prices"][m["Product index"]]

# ACCESS THE TABLE
product_id_table = session.tables["ProductID"]
# YOU CAN MANIPULATE THE TABLE USING `load_csv()` for instance
product_id_table.load_csv(...)
```
The hierarchy is updated when the table is updated so the goal is to update the Table when the csv is updated.

For this I see some solutions:
1. A file watcher like in [this how to](https://docs.atoti.io/latest/how_tos/watch/file_watching.html)
2. If it is updated at a specific hour you can create a job triggering the loading
3. Maybe use an [`endpoint()`](https://docs.atoti.io/latest/lib/atoti/atoti.session.html#atoti.Session.endpoint)

**Comment by EricDallAgnol** [comment]
Hello,

Thanks for you answer and the solutions !
Ok good to know a table is created under the hood !
